### PR TITLE
Import TabletState and remove mutex in TabletMeta

### PR DESCRIPTION
**Pull request by chaoyli**
1.  Import TabletState to describe tablet state transitions.
2. There is no necessary to use std::mutex in TabletMeta,
     because concurrency conflict addressed by Tablet.

### CODE REVIEW

File: be/src/olap/tablet_manager.cpp [hunk 1]
```diff
@@ -555,8 +552,8 @@ TabletSharedPtr TabletManager::get_tablet(TTabletId tablet_id, SchemaHash schema
         if (!tablet->is_used()) {
             OLAP_LOG_WARNING("tablet cannot be used. [tablet=%ld]", tablet_id);
             tablet.reset();
-        } else if (load_tablet && !tablet->is_loaded()) {
-            if (tablet->load() != OLAP_SUCCESS) {
+        } else if (load_tablet && !tablet->init_success()) {
+            if (tablet->init() != OLAP_SUCCESS) {
                 OLAP_LOG_WARNING("fail to load tablet. [tablet=%ld]", tablet_id);
```
Comment (kangpinghuang) [inline be/src/olap/tablet_manager.cpp]: LOG(INFO)

File: be/src/olap/tablet_manager.cpp [hunk 2]
```diff
@@ -555,8 +552,8 @@ TabletSharedPtr TabletManager::get_tablet(TTabletId tablet_id, SchemaHash schema
         if (!tablet->is_used()) {
             OLAP_LOG_WARNING("tablet cannot be used. [tablet=%ld]", tablet_id);
             tablet.reset();
-        } else if (load_tablet && !tablet->is_loaded()) {
-            if (tablet->load() != OLAP_SUCCESS) {
+        } else if (load_tablet && !tablet->init_success()) {
```
Comment (kangpinghuang) [inline be/src/olap/tablet_manager.cpp]: I think is_inited is a better name?

File: be/src/olap/tablet.h [hunk 3]
```diff
@@ -27,221 +27,211 @@
...
-    OLAPStatus test_version(const Version& version);
-    size_t get_version_data_size(const Version& version);
-    OLAPStatus recover_tablet_until_specfic_version(const int64_t& spec_version,
-                                                    const int64_t& version_hash);
-    const std::string& rowset_path_prefix();
-    const size_t id() { return _id; }
-    void set_id(size_t id) { _id = id; }
+    inline DataDir* data_dir() const;
     OLAPStatus register_tablet_into_dir();
+    std::string dir_path() const;
```
Comment (kangpinghuang) [inline be/src/olap/tablet.h]: I think dir_path is not a good name? data_dir_path?

File: be/src/olap/tablet.cpp [hunk 4]
```diff
@@ -73,8 +56,8 @@ TabletSharedPtr Tablet::create_from_tablet_meta_file(
 
     // add new fields
     boost::filesystem::path file_path_path(file_path);
-    std::string shard_path = file_path_path.parent_path().parent_path().parent_path().string();
-    std::string shard_str = shard_path.substr(shard_path.find_last_of('/') + 1);
+    string shard_path = file_path_path.parent_path().parent_path().parent_path().string();
+    string shard_str = shard_path.substr(shard_path.find_last_of('/') + 1);
```
Comment (kangpinghuang) [inline be/src/olap/tablet.cpp]: you delete using std::string, can this be compiled?

File: be/src/olap/base_compaction.cpp [hunk 5]
```diff
@@ -200,7 +200,7 @@ bool BaseCompaction::_check_whether_satisfy_policy(bool is_manual_trigger,
         // base文件
         if (temp.first == 0) {
             _old_base_version = temp;
-            base_size = _tablet->get_version_data_size(temp);
+            base_size = _tablet->get_rowset_size_by_version(temp);
```
Comment (kangpinghuang) [inline be/src/olap/base_compaction.cpp]: temp is not a good name?

File: be/src/olap/base_compaction.cpp [hunk 6]
```diff
@@ -43,7 +43,7 @@ namespace doris {
 OLAPStatus BaseCompaction::init(TabletSharedPtr tablet, bool is_manual_trigger) {
     // 表在首次查询或PUSH等操作时，会被加载到内存
     // 如果表没有被加载，表明该表上目前没有任何操作，所以不进行BE操作
-    if (!tablet->is_loaded()) {
+    if (!tablet->init_success()) {
```
Comment (kangpinghuang) [inline be/src/olap/base_compaction.cpp]: init_success -> is_inited?

**Review by kangpinghuang — CHANGES_REQUESTED** [review]

File: be/src/olap/tablet_meta.h [hunk 7]
```diff
@@ -91,29 +112,51 @@ class TabletMeta {
...
+    OLAPStatus init_from_pb(const TabletMetaPB& tablet_meta_pb);
 
-    OLAPStatus save_meta();
-    OLAPStatus save_meta_unlock();
+    OLAPStatus to_meta_pb(TabletMetaPB* tablet_meta_pb);
+    OLAPStatus to_json(std::string* json_string, json2pb::Pb2JsonOptions& options);
 
-    OLAPStatus to_tablet_pb(TabletMetaPB* tablet_meta_pb);
-    OLAPStatus to_tablet_pb_unlock(TabletMetaPB* tablet_meta_pb);
+    inline void set_data_dir(DataDir* data_dir);
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/tablet_meta.h]: When cloning tablet_meta from remote host, data_dir is empty.

File: be/src/olap/tablet_meta.h [hunk 8]
```diff
@@ -123,80 +166,55 @@ class TabletMeta {
...
     DelPredicateArray _del_pred_array;
 
-    TabletState _tablet_state;
     AlterTabletTask _alter_task;
 
-    TabletMetaPB _tablet_meta_pb;
     DataDir* _data_dir;
-
-    mutable std::mutex _mutex;
+    TabletMetaPB _tablet_meta_pb;
```
[the remark this answers is not in the document]
Comment (chaoyli) [reply be/src/olap/tablet_meta.h]: In construtor

File: be/src/olap/tablet_manager.cpp (hunk 1 above)
Comment (chaoyli) [reply be/src/olap/tablet_manager.cpp]: OK

File: be/src/olap/tablet.h (hunk 3 above)
Comment (chaoyli) [reply be/src/olap/tablet.h]: OK, I remove dir_path() from tablet.h

File: be/src/olap/tablet.cpp (hunk 4 above)
Comment (chaoyli) [reply be/src/olap/tablet.cpp]: using std::string in tablet.h

File: be/src/olap/tablet_manager.cpp (hunk 2 above)
Comment (chaoyli) [reply be/src/olap/tablet_manager.cpp]: I think init_succeeded may be better to describe it.
is_inited can not differentiate success or fail literally

**Review by kangpinghuang — APPROVED** [review]

**Pull request merged**